File: src/main/cpp/subsystems/climber/ClimberIOSim.cpp
// Copyright (c) 2025 FRC 6328
// http://github.com/Mechanical-Advantage
//
// Use of this source code is governed by an MIT-style
// license that can be found in the LICENSE file at
// the root directory of this project.

#include "ClimberIOSim.h"
#include "frc/MathUtil.h"
#include "frc/geometry/Rotation2d.h"
#include "frc/math/Matrix.h"
#include "frc/math/VecBuilder.h"
#include "frc/math/numbers/N1.h"
#include "frc/math/numbers/N2.h"
#include "frc/math/system/NumericalIntegration.h"
#include "frc/math/system/plant/DCMotor.h"
#include "org/littletonrobotics/frc2025/Constants.h"
#include "org/littletonrobotics/frc2025/subsystems/superstructure/SuperstructureConstants.h"
#include "org/littletonrobotics/frc2025/subsystems/superstructure/SuperstructureVisualizer.h"

ClimberIOSim::ClimberIOSim() {
  simState = frc::VecBuilder<frc::numbers::N2>::Fill(wpi::math::pi / 2.0, 0.0);
}

void ClimberIOSim::UpdateInputs(ClimberIOInputs &inputs) {
  Update(org::littletonrobotics::frc2025::Constants::loopPeriodSecs);
  org::littletonrobotics::frc2025::subsystems::superstructure::
      SuperstructureVisualizer::UpdateSimIntake(simState(0));

  inputs.data.motorConnected = true;
  inputs.data.positionRads = units::radian_t(simState(0));
  inputs.data.velocityRadsPerSec = units::radians_per_second_t(simState(1));
  inputs.data.appliedVoltage = units::volt_t(appliedVolts);
  inputs.data.torqueCurrentAmps =
      units::ampere_t(wpi::math::copySign(inputTorqueCurrent, appliedVolts));
  inputs.data.supplyCurrentAmps =
      units::ampere_t(wpi::math::copySign(inputTorqueCurrent, appliedVolts));
  inputs.data.tempCelsius = units::degree_celsius_t(0.0);
}

void ClimberIOSim::RunTorqueCurrent(double current) {
  inputTorqueCurrent = current;
  appliedVolts =
      gearbox.GetVoltage(gearbox.GetTorque(inputTorqueCurrent), simState(1));
  appliedVolts = frc::math::clamp(appliedVolts, -12.0, 12.0);
}

void ClimberIOSim::Stop() { RunTorqueCurrent(0.0); }

void ClimberIOSim::Update(double dt) {
  inputTorqueCurrent = frc::math::clamp(
      inputTorqueCurrent, -gearbox.StallCurrent(), gearbox.StallCurrent());
  frc::Matrix<frc::numbers::N2, frc::numbers::N1> updatedState =
      frc::NumericalIntegration::RKDP(
          [](const frc::Matrix<frc::numbers::N2, frc::numbers::N1> &x,
             const frc::Matrix<frc::numbers::N1, frc::numbers::N1> &u) {
            return A * x + B * u +
                   (-org::littletonrobotics::frc2025::subsystems::
                        superstructure::SuperstructureConstants::G *
                    cgRadius * frc::Rotation2d(units::radian_t(x(0))).Cos() /
                    moi);
          },
          simState,
          frc::VecBuilder<frc::numbers::N1>::Fill(
              inputTorqueCurrent * 15.0), // Magic constant of doom
          dt);
  simState =
      frc::VecBuilder<frc::numbers::N2>::Fill(updatedState(0), updatedState(1));
}

File: src/main/java/org/littletonrobotics/frc2025/subsystems/climber/ClimberIOSim.java
// Copyright (c) 2025 FRC 6328
// http://github.com/Mechanical-Advantage
//
// Use of this source code is governed by an MIT-style
// license that can be found in the LICENSE file at
// the root directory of this project.

package org.littletonrobotics.frc2025.subsystems.climber;

import edu.wpi.first.math.*;
import edu.wpi.first.math.numbers.N1;
import edu.wpi.first.math.numbers.N2;
import edu.wpi.first.math.system.NumericalIntegration;
import edu.wpi.first.math.system.plant.DCMotor;
import org.littletonrobotics.frc2025.Constants;

public class ClimberIOSim implements ClimberIO {
  private static final double moi = 1.2;
  private static final DCMotor gearbox =
      DCMotor.getKrakenX60Foc(1).withReduction(ClimberIOTalonFX.reduction);
  private static final Matrix<N2, N2> A =
      MatBuilder.fill(
          Nat.N2(),
          Nat.N2(),
          0,
          1,
          0,
          -gearbox.KtNMPerAmp / (gearbox.KvRadPerSecPerVolt * gearbox.rOhms * moi));
  private static final Vector<N2> B = VecBuilder.fill(0, gearbox.KtNMPerAmp / moi);

  // Climber sim
  private Vector<N2> simState;
  private double inputTorqueCurrent = 0.0;
  private double appliedVolts = 0.0;

  public ClimberIOSim() {
    simState = VecBuilder.fill(0.0, 0.0);
  }

  @Override
  public void updateInputs(ClimberIOInputs inputs) {
    update(Constants.loopPeriodSecs);

    inputs.data =
        new ClimberIOData(
            true,
            simState.get(0),
            simState.get(1),
            appliedVolts,
            Math.copySign(inputTorqueCurrent, appliedVolts),
            Math.copySign(inputTorqueCurrent, appliedVolts),
            0.0);
  }

  @Override
  public void runTorqueCurrent(double current) {
    inputTorqueCurrent = current;
    appliedVolts = gearbox.getVoltage(gearbox.getTorque(inputTorqueCurrent), simState.get(1, 0));
    appliedVolts = MathUtil.clamp(appliedVolts, -12.0, 12.0);
  }

  @Override
  public void stop() {
    runTorqueCurrent(0.0);
  }

  private void update(double dt) {
    inputTorqueCurrent =
        MathUtil.clamp(inputTorqueCurrent, -gearbox.stallCurrentAmps, gearbox.stallCurrentAmps);
    Matrix<N2, N1> updatedState =
        NumericalIntegration.rkdp(
            (Matrix<N2, N1> x, Matrix<N1, N1> u) -> A.times(x).plus(B.times(u)),
            simState,
            VecBuilder.fill(inputTorqueCurrent * 15), // Magic constant of doom
            dt);
    simState = VecBuilder.fill(updatedState.get(0, 0), updatedState.get(1, 0));
  }
}
